File: scholarsphere_backend/app/schemas/testimonial.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.testimonial import (
    TestimonialDisplayMode,
    TestimonialOutcome,
    TestimonialReactionType,
    TestimonialStatus,
    TestimonialVerificationStatus,
)

#: A closed vocabulary rather than free text, so the public "Features
#: used" filter/badges (Phase 4) can't drift into one-off strings that
#: never match anything. Mirrors the platform's real feature surface.
KNOWN_FEATURES: tuple[str, ...] = (
    "opportunity_discovery",
    "eligibility_matching",
    "application_preparation",
    "cv_builder",
    "document_management",
    "deadline_tracking",
    "guidance_plans",
    "ai_application_assistant",
    "calendar_sync",
    "notifications",
)

#: Character caps on free-text story fields. Generous enough for a real
#: case study, but bounded so a single submission can't exhaust storage
#: or make the moderation queue unreadable.
_SHORT_TEXT_MAX = 4000
_NAME_MAX = 255


def display_name_for(
    full_name: str, display_mode: TestimonialDisplayMode
) -> str:
    """Compute the *shown* name from the stored real name - never stored
    separately, so changing `display_mode` later can't leak a name a
    previous read already cached under a different mode.
    """
    name = full_name.strip()
    if display_mode == TestimonialDisplayMode.anonymous or not name:
        return "Anonymous Applicant"
    if display_mode == TestimonialDisplayMode.full_name:
        return name
    parts = name.split()
    if len(parts) == 1:
        return parts[0]
    first, last = parts[0], parts[-1]
    return f"{first} {last[0]}." if last else first


class TestimonialOpportunityInput(BaseModel):
    opportunity_id: UUID | None = None
    opportunity_name: str = Field(min_length=1, max_length=500)
    opportunity_provider: str = Field(min_length=1, max_length=512)
    opportunity_type: str = Field(min_length=1, max_length=128)
    country: str | None = Field(default=None, max_length=255)
    degree_level: str | None = Field(default=None, max_length=128)
    field_of_study: str | None = Field(default=None, max_length=255)
    success_year: int | None = Field(default=None, ge=1990, le=2100)
    outcome: TestimonialOutcome


class TestimonialExperienceInput(BaseModel):
    challenge: str | None = Field(default=None, max_length=_SHORT_TEXT_MAX)
    discovery_story: str | None = Field(default=None, max_length=_SHORT_TEXT_MAX)
    preparation_story: str | None = Field(default=None, max_length=_SHORT_TEXT_MAX)
    scholarsphere_help: str | None = Field(default=None, max_length=_SHORT_TEXT_MAX)
    outcome_narrative: str | None = Field(default=None, max_length=_SHORT_TEXT_MAX)
    impact: str | None = Field(default=None, max_length=_SHORT_TEXT_MAX)
    advice: str | None = Field(default=None, max_length=_SHORT_TEXT_MAX)
    features_used: list[str] = Field(default_factory=list)

    @field_validator("features_used")
    @classmethod
    def _known_features_only(cls, value: list[str]) -> list[str]:
        unknown = [item for item in value if item not in KNOWN_FEATURES]
        if unknown:
            raise ValueError(f"Unknown feature(s): {unknown!r}")
        # De-duplicate while preserving order.
        return list(dict.fromkeys(value))


class TestimonialProfileInput(BaseModel):
    full_name: str = Field(min_length=1, max_length=_NAME_MAX)
    university: str | None = Field(default=None, max_length=255)
    program: str | None = Field(default=None, max_length=255)
    photo_storage_path: str | None = Field(default=None, max_length=1024)


class TestimonialPrivacyInput(BaseModel):
    display_mode: TestimonialDisplayMode = TestimonialDisplayMode.first_name_last_initial
    show_university: bool = True
    show_country: bool = True
    show_program: bool = True
    show_photo: bool = False


class TestimonialDraftRequest(BaseModel):
    """The full submission wizard's payload - every step's fields in one
    body. Steps 1-5 map directly onto the four input models above; step 6
    (review) is client-side only (a preview of this same data); step 7
    (consent) is `consent_confirmed`, only meaningful at submit time.
    """

    opportunity: TestimonialOpportunityInput
    experience: TestimonialExperienceInput = TestimonialExperienceInput()
    profile: TestimonialProfileInput
    privacy: TestimonialPrivacyInput = TestimonialPrivacyInput()
    evidence_storage_paths: list[str] = Field(default_factory=list, max_length=5)


class TestimonialSubmitRequest(BaseModel):
    consent_confirmed: bool


class TestimonialModerationDecisionRequest(BaseModel):
    reason: str = Field(min_length=1, max_length=_SHORT_TEXT_MAX)


class TestimonialVerifyRequest(BaseModel):
    verification_method: str = Field(min_length=1, max_length=255)
    notes: str | None = Field(default=None, max_length=_SHORT_TEXT_MAX)


class TestimonialReactionRequest(BaseModel):
    reaction_type: TestimonialReactionType


class TestimonialInternalNotesRequest(BaseModel):
    """Staff-only working notes - never serialized to `TestimonialOwnRead`
    or any public/summary schema, only `TestimonialAdminRead` (Phase 33:
    "Never expose internal verification notes publicly").
    """

    notes: str = Field(default="", max_length=_SHORT_TEXT_MAX)


def _badges(status: TestimonialStatus, verification_status: TestimonialVerificationStatus, featured: bool) -> list[str]:
    badges: list[str] = []
    if featured and status == TestimonialStatus.approved:
        badges.append("featured")
    if verification_status == TestimonialVerificationStatus.verified:
        badges.append("verified")
    elif status == TestimonialStatus.approved:
        badges.append("community")
    else:
        badges.append("under_review")
    return badges


class TestimonialSummaryRead(BaseModel):
    """Card-shaped view for list/search/featured/related-stories surfaces.

    Never includes `full_name`, `evidence_storage_paths`,
    `internal_notes`, or `rejection_reason` - a summary read is
    reachable by any authenticated user, not just the owner or staff.
    """

    id: UUID
    slug: str
    display_name: str
    photo_storage_path: str | None
    country: str | None
    university: str | None
    program: str | None
    degree_level: str | None
    field_of_study: str | None
    opportunity_name: str
    opportunity_provider: str
    opportunity_type: str
    outcome: TestimonialOutcome
    success_year: int | None
    verification_status: TestimonialVerificationStatus
    featured: bool
    badges: list[str]
    excerpt: str
    features_used: list[str]
    helpful_count: int
    inspiring_count: int
    useful_count: int
    created_at: datetime

    @classmethod
    def from_model(cls, row: Any) -> "TestimonialSummaryRead":
        display_name = display_name_for(row.full_name, row.display_mode)
        excerpt_source = row.outcome_narrative or row.scholarsphere_help or row.impact or ""
        excerpt = excerpt_source.strip()
        if len(excerpt) > 220:
            excerpt = excerpt[:217].rstrip() + "..."
        return cls(
            id=row.id,
            slug=row.slug,
            display_name=display_name,
            photo_storage_path=row.photo_storage_path if row.show_photo else None,
            country=row.country if row.show_country else None,
            university=row.university if row.show_university else None,
            program=row.program if row.show_program else None,
            degree_level=row.degree_level,
            field_of_study=row.field_of_study,
            opportunity_name=row.opportunity_name,
            opportunity_provider=row.opportunity_provider,
            opportunity_type=row.opportunity_type,
            outcome=row.outcome,
            success_year=row.success_year,
            verification_status=row.verification_status,
            featured=row.featured,
            badges=_badges(row.status, row.verification_status, row.featured),
            excerpt=excerpt,
            features_used=list(row.features_used),
            helpful_count=row.helpful_count,
            inspiring_count=row.inspiring_count,
            useful_count=row.useful_count,
            created_at=row.created_at,
        )


class TestimonialDetailRead(BaseModel):
    """Full public story-page view. Same privacy exclusions as the
    summary above, plus the long-form narrative fields.
    """

    id: UUID
    slug: str
    display_name: str
    photo_storage_path: str | None
    country: str | None
    university: str | None
    program: str | None
    degree_level: str | None
    field_of_study: str | None
    opportunity_id: UUID | None
    opportunity_name: str
    opportunity_provider: str
    opportunity_type: str
    outcome: TestimonialOutcome
    success_year: int | None
    challenge: str | None
    discovery_story: str | None
    preparation_story: str | None
    scholarsphere_help: str | None
    outcome_narrative: str | None
    impact: str | None
    advice: str | None
    features_used: list[str]
    verification_status: TestimonialVerificationStatus
    featured: bool
    badges: list[str]
    helpful_count: int
    inspiring_count: int
    useful_count: int
    view_count: int
    created_at: datetime

    @classmethod
    def from_model(cls, row: Any) -> "TestimonialDetailRead":
        return cls(
            id=row.id,
            slug=row.slug,
            display_name=display_name_for(row.full_name, row.display_mode),
            photo_storage_path=row.photo_storage_path if row.show_photo else None,
            country=row.country if row.show_country else None,
            university=row.university if row.show_university else None,
            program=row.program if row.show_program else None,
            degree_level=row.degree_level,
            field_of_study=row.field_of_study,
            opportunity_id=row.opportunity_id,
            opportunity_name=row.opportunity_name,
            opportunity_provider=row.opportunity_provider,
            opportunity_type=row.opportunity_type,
            outcome=row.outcome,
            success_year=row.success_year,
            challenge=row.challenge,
            discovery_story=row.discovery_story,
            preparation_story=row.preparation_story,
            scholarsphere_help=row.scholarsphere_help,
            outcome_narrative=row.outcome_narrative,
            impact=row.impact,
            advice=row.advice,
            features_used=list(row.features_used),
            verification_status=row.verification_status,
            featured=row.featured,
            badges=_badges(row.status, row.verification_status, row.featured),
            helpful_count=row.helpful_count,
            inspiring_count=row.inspiring_count,
            useful_count=row.useful_count,
            view_count=row.view_count,
            created_at=row.created_at,
        )


class TestimonialOwnRead(BaseModel):
    """The owner's own view of their submission - unlike the public
    reads above, this always shows the real (unmasked) fields the owner
    themselves entered, plus editorial state, so they can see exactly
    what was submitted and why it might have been rejected.
    Deliberately still excludes `internal_notes` (staff-only).
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    slug: str
    status: TestimonialStatus
    verification_status: TestimonialVerificationStatus
    featured: bool
    opportunity_id: UUID | None
    opportunity_name: str
    opportunity_provider: str
    opportunity_type: str
    country: str | None
    degree_level: str | None
    field_of_study: str | None
    success_year: int | None
    outcome: TestimonialOutcome
    challenge: str | None
    discovery_story: str | None
    preparation_story: str | None
    scholarsphere_help: str | None
    outcome_narrative: str | None
    impact: str | None
    advice: str | None
    features_used: list[str]
    full_name: str
    university: str | None
    program: str | None
    photo_storage_path: str | None
    display_mode: TestimonialDisplayMode
    show_university: bool
    show_country: bool
    show_program: bool
    show_photo: bool
    evidence_storage_paths: list[str]
    rejection_reason: str | None
    submitted_at: datetime | None
    approved_at: datetime | None
    verified_at: datetime | None
    withdrawn_at: datetime | None
    created_at: datetime
    updated_at: datetime


class TestimonialAdminRead(TestimonialOwnRead):
    """Everything the owner sees, plus staff-only fields."""

    user_id: str
    internal_notes: str | None
    verified_by: str | None
    verification_method: str | None
    last_moderator_id: str | None
    rejected_at: datetime | None


class TestimonialModerationHistoryItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    previous_status: str
    new_status: str
    actor_id: str
    notes: str
    created_at: datetime


class TestimonialPage(BaseModel):
    items: list[TestimonialSummaryRead]
    total: int
    page: int
    page_size: int


class TestimonialAdminPage(BaseModel):
    items: list[TestimonialAdminRead]
    total: int
    page: int
    page_size: int


class TestimonialStatsRead(BaseModel):
    """Every count here is a real aggregate query against `testimonials`
    at request time - never a fabricated or cached-forever number (Phase
    3's "never fabricate numbers" requirement). A count of zero is
    represented as `None` by the service for a field the UI should hide
    rather than show as "0", per Phase 3's "gracefully hide the
    statistic rather than displaying fake numbers".
    """

    total_stories: int | None
    verified_stories: int | None
    countries_represented: int | None
    opportunity_types_represented: int | None
    fields_of_study_represented: int | None

File: lib/features/testimonials/domain/testimonial.dart
/// Mirrors scholarsphere_backend/app/models/testimonial.py and
/// app/schemas/testimonial.py exactly - the wire values below (e.g.
/// `awarded`, `first_name_last_initial`) are the backend's own enum
/// string values, never re-invented client-side.
library;

enum TestimonialOutcome {
  applied,
  shortlisted,
  interviewed,
  selected,
  awarded,
  admitted,
  funded,
  other;

  static TestimonialOutcome fromWire(String value) => values.firstWhere(
    (item) => item.wire == value,
    orElse: () => TestimonialOutcome.other,
  );

  String get wire => name;

  String get label => switch (this) {
    TestimonialOutcome.applied => 'Applied',
    TestimonialOutcome.shortlisted => 'Shortlisted',
    TestimonialOutcome.interviewed => 'Interviewed',
    TestimonialOutcome.selected => 'Selected',
    TestimonialOutcome.awarded => 'Awarded',
    TestimonialOutcome.admitted => 'Admitted',
    TestimonialOutcome.funded => 'Funded',
    TestimonialOutcome.other => 'Other outcome',
  };
}

enum TestimonialStatus {
  draft,
  submitted,
  underReview,
  changesRequested,
  approved,
  rejected,
  withdrawn;

  static TestimonialStatus fromWire(String value) => switch (value) {
    'draft' => TestimonialStatus.draft,
    'submitted' => TestimonialStatus.submitted,
    'under_review' => TestimonialStatus.underReview,
    'changes_requested' => TestimonialStatus.changesRequested,
    'approved' => TestimonialStatus.approved,
    'rejected' => TestimonialStatus.rejected,
    'withdrawn' => TestimonialStatus.withdrawn,
    _ => TestimonialStatus.draft,
  };

  String get label => switch (this) {
    TestimonialStatus.draft => 'Draft',
    TestimonialStatus.submitted => 'Submitted',
    TestimonialStatus.underReview => 'Under review',
    TestimonialStatus.changesRequested => 'Changes requested',
    TestimonialStatus.approved => 'Published',
    TestimonialStatus.rejected => 'Not approved',
    TestimonialStatus.withdrawn => 'Withdrawn',
  };
}

enum TestimonialVerificationStatus {
  unverified,
  verified;

  static TestimonialVerificationStatus fromWire(String value) =>
      value == 'verified'
      ? TestimonialVerificationStatus.verified
      : TestimonialVerificationStatus.unverified;
}

enum TestimonialDisplayMode {
  fullName,
  firstNameLastInitial,
  anonymous;

  static TestimonialDisplayMode fromWire(String value) => switch (value) {
    'full_name' => TestimonialDisplayMode.fullName,
    'anonymous' => TestimonialDisplayMode.anonymous,
    _ => TestimonialDisplayMode.firstNameLastInitial,
  };

  String get wire => switch (this) {
    TestimonialDisplayMode.fullName => 'full_name',
    TestimonialDisplayMode.firstNameLastInitial => 'first_name_last_initial',
    TestimonialDisplayMode.anonymous => 'anonymous',
  };

  String get label => switch (this) {
    TestimonialDisplayMode.fullName => 'Full name',
    TestimonialDisplayMode.firstNameLastInitial => 'First name + last initial',
    TestimonialDisplayMode.anonymous => 'Anonymous',
  };
}

enum TestimonialReactionType {
  helpful,
  inspiring,
  useful;

  String get wire => name;

  String get label => switch (this) {
    TestimonialReactionType.helpful => 'Helpful',
    TestimonialReactionType.inspiring => 'Inspiring',
    TestimonialReactionType.useful => 'Useful',
  };
}

/// A closed vocabulary matching KNOWN_FEATURES in
/// app/schemas/testimonial.py - the submission form only lets applicants
/// pick from this list, so "features used" badges are always meaningful.
const List<String> knownTestimonialFeatures = [
  'opportunity_discovery',
  'eligibility_matching',
  'application_preparation',
  'cv_builder',
  'document_management',
  'deadline_tracking',
  'guidance_plans',
  'ai_application_assistant',
  'calendar_sync',
  'notifications',
];

String featureLabel(String key) => switch (key) {
  'opportunity_discovery' => 'Opportunity Discovery',
  'eligibility_matching' => 'Eligibility Matching',
  'application_preparation' => 'Application Preparation',
  'cv_builder' => 'CV Builder',
  'document_management' => 'Document Management',
  'deadline_tracking' => 'Deadline Tracking',
  'guidance_plans' => 'Guidance Plans',
  'ai_application_assistant' => 'AI Application Assistant',
  'calendar_sync' => 'Calendar Sync',
  'notifications' => 'Notifications',
  _ => key,
};

/// Card-shaped view for list/search/featured/related-stories surfaces -
/// mirrors TestimonialSummaryRead. Never carries the real name, evidence,
/// or any other field the owner might have chosen to withhold.
class SuccessStorySummary {
  const SuccessStorySummary({
    required this.id,
    required this.slug,
    required this.displayName,
    required this.photoStoragePath,
    required this.country,
    required this.university,
    required this.program,
    required this.degreeLevel,
    required this.fieldOfStudy,
    required this.opportunityName,
    required this.opportunityProvider,
    required this.opportunityType,
    required this.outcome,
    required this.successYear,
    required this.verificationStatus,
    required this.featured,
    required this.badges,
    required this.excerpt,
    required this.featuresUsed,
    required this.helpfulCount,
    required this.inspiringCount,
    required this.usefulCount,
    required this.createdAt,
  });

  final String id;
  final String slug;
  final String displayName;
  final String? photoStoragePath;
  final String? country;
  final String? university;
  final String? program;
  final String? degreeLevel;
  final String? fieldOfStudy;
  final String opportunityName;
  final String opportunityProvider;
  final String opportunityType;
  final TestimonialOutcome outcome;
  final int? successYear;
  final TestimonialVerificationStatus verificationStatus;
  final bool featured;
  final List<String> badges;
  final String excerpt;
  final List<String> featuresUsed;
  final int helpfulCount;
  final int inspiringCount;
  final int usefulCount;
  final DateTime createdAt;

  bool get isVerified =>
      verificationStatus == TestimonialVerificationStatus.verified;
}

class SuccessStoryDetail {
  const SuccessStoryDetail({
    required this.id,
    required this.slug,
    required this.displayName,
    required this.photoStoragePath,
    required this.country,
    required this.university,
    required this.program,
    required this.degreeLevel,
    required this.fieldOfStudy,
    required this.opportunityId,
    required this.opportunityName,
    required this.opportunityProvider,
    required this.opportunityType,
    required this.outcome,
    required this.successYear,
    required this.challenge,
    required this.discoveryStory,
    required this.preparationStory,
    required this.scholarsphereHelp,
    required this.outcomeNarrative,
    required this.impact,
    required this.advice,
    required this.featuresUsed,
    required this.verificationStatus,
    required this.featured,
    required this.badges,
    required this.helpfulCount,
    required this.inspiringCount,
    required this.usefulCount,
    required this.viewCount,
    required this.createdAt,
  });

  final String id;
  final String slug;
  final String displayName;
  final String? photoStoragePath;
  final String? country;
  final String? university;
  final String? program;
  final String? degreeLevel;
  final String? fieldOfStudy;
  final String? opportunityId;
  final String opportunityName;
  final String opportunityProvider;
  final String opportunityType;
  final TestimonialOutcome outcome;
  final int? successYear;
  final String? challenge;
  final String? discoveryStory;
  final String? preparationStory;
  final String? scholarsphereHelp;
  final String? outcomeNarrative;
  final String? impact;
  final String? advice;
  final List<String> featuresUsed;
  final TestimonialVerificationStatus verificationStatus;
  final bool featured;
  final List<String> badges;
  final int helpfulCount;
  final int inspiringCount;
  final int usefulCount;
  final int viewCount;
  final DateTime createdAt;

  bool get isVerified =>
      verificationStatus == TestimonialVerificationStatus.verified;
}

/// The owner's own view of a submission, at any status - always shows the
/// real fields they entered (mirrors TestimonialOwnRead).
class MyTestimonial {
  const MyTestimonial({
    required this.id,
    required this.slug,
    required this.status,
    required this.verificationStatus,
    required this.featured,
    required this.opportunityId,
    required this.opportunityName,
    required this.opportunityProvider,
    required this.opportunityType,
    required this.country,
    required this.degreeLevel,
    required this.fieldOfStudy,
    required this.successYear,
    required this.outcome,
    required this.challenge,
    required this.discoveryStory,
    required this.preparationStory,
    required this.scholarsphereHelp,
    required this.outcomeNarrative,
    required this.impact,
    required this.advice,
    required this.featuresUsed,
    required this.fullName,
    required this.university,
    required this.program,
    required this.photoStoragePath,
    required this.displayMode,
    required this.showUniversity,
    required this.showCountry,
    required this.showProgram,
    required this.showPhoto,
    required this.evidenceStoragePaths,
    required this.rejectionReason,
    required this.submittedAt,
    required this.approvedAt,
    required this.verifiedAt,
    required this.withdrawnAt,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String slug;
  final TestimonialStatus status;
  final TestimonialVerificationStatus verificationStatus;
  final bool featured;
  final String? opportunityId;
  final String opportunityName;
  final String opportunityProvider;
  final String opportunityType;
  final String? country;
  final String? degreeLevel;
  final String? fieldOfStudy;
  final int? successYear;
  final TestimonialOutcome outcome;
  final String? challenge;
  final String? discoveryStory;
  final String? preparationStory;
  final String? scholarsphereHelp;
  final String? outcomeNarrative;
  final String? impact;
  final String? advice;
  final List<String> featuresUsed;
  final String fullName;
  final String? university;
  final String? program;
  final String? photoStoragePath;
  final TestimonialDisplayMode displayMode;
  final bool showUniversity;
  final bool showCountry;
  final bool showProgram;
  final bool showPhoto;
  final List<String> evidenceStoragePaths;
  final String? rejectionReason;
  final DateTime? submittedAt;
  final DateTime? approvedAt;
  final DateTime? verifiedAt;
  final DateTime? withdrawnAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  bool get isEditable =>
      status == TestimonialStatus.draft ||
      status == TestimonialStatus.changesRequested;
  bool get isWithdrawable =>
      status == TestimonialStatus.submitted ||
      status == TestimonialStatus.underReview ||
      status == TestimonialStatus.approved;
}

class AdminTestimonial extends MyTestimonial {
  const AdminTestimonial({
    required super.id,
    required super.slug,
    required super.status,
    required super.verificationStatus,
    required super.featured,
    required super.opportunityId,
    required super.opportunityName,
    required super.opportunityProvider,
    required super.opportunityType,
    required super.country,
    required super.degreeLevel,
    required super.fieldOfStudy,
    required super.successYear,
    required super.outcome,
    required super.challenge,
    required super.discoveryStory,
    required super.preparationStory,
    required super.scholarsphereHelp,
    required super.outcomeNarrative,
    required super.impact,
    required super.advice,
    required super.featuresUsed,
    required super.fullName,
    required super.university,
    required super.program,
    required super.photoStoragePath,
    required super.displayMode,
    required super.showUniversity,
    required super.showCountry,
    required super.showProgram,
    required super.showPhoto,
    required super.evidenceStoragePaths,
    required super.rejectionReason,
    required super.submittedAt,
    required super.approvedAt,
    required super.verifiedAt,
    required super.withdrawnAt,
    required super.createdAt,
    required super.updatedAt,
    required this.userId,
    required this.internalNotes,
    required this.verifiedBy,
    required this.verificationMethod,
    required this.lastModeratorId,
    required this.rejectedAt,
  });

  final String userId;
  final String? internalNotes;
  final String? verifiedBy;
  final String? verificationMethod;
  final String? lastModeratorId;
  final DateTime? rejectedAt;
}

class TestimonialModerationHistoryItem {
  const TestimonialModerationHistoryItem({
    required this.id,
    required this.previousStatus,
    required this.newStatus,
    required this.actorId,
    required this.notes,
    required this.createdAt,
  });

  final String id;
  final String previousStatus;
  final String newStatus;
  final String actorId;
  final String notes;
  final DateTime createdAt;
}

class SuccessStoryPage {
  const SuccessStoryPage({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  final List<SuccessStorySummary> items;
  final int total;
  final int page;
  final int pageSize;

  bool get hasMore => page * pageSize < total;
}

class AdminTestimonialPage {
  const AdminTestimonialPage({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  final List<AdminTestimonial> items;
  final int total;
  final int page;
  final int pageSize;

  bool get hasMore => page * pageSize < total;
}

/// Every field is nullable - a `null` count means "hide this statistic",
/// never a fabricated zero (mirrors TestimonialStatsRead's own contract).
class TestimonialStats {
  const TestimonialStats({
    required this.totalStories,
    required this.verifiedStories,
    required this.countriesRepresented,
    required this.opportunityTypesRepresented,
    required this.fieldsOfStudyRepresented,
  });

  final int? totalStories;
  final int? verifiedStories;
  final int? countriesRepresented;
  final int? opportunityTypesRepresented;
  final int? fieldsOfStudyRepresented;

  bool get hasAnyStat =>
      totalStories != null ||
      verifiedStories != null ||
      countriesRepresented != null ||
      opportunityTypesRepresented != null ||
      fieldsOfStudyRepresented != null;
}

/// The full submission-wizard payload - steps 1-5 of the form map
/// directly onto these fields, mirroring TestimonialDraftRequest.
class TestimonialDraftInput {
  const TestimonialDraftInput({
    this.opportunityId,
    required this.opportunityName,
    required this.opportunityProvider,
    required this.opportunityType,
    this.country,
    this.degreeLevel,
    this.fieldOfStudy,
    this.successYear,
    required this.outcome,
    this.challenge,
    this.discoveryStory,
    this.preparationStory,
    this.scholarsphereHelp,
    this.outcomeNarrative,
    this.impact,
    this.advice,
    this.featuresUsed = const [],
    required this.fullName,
    this.university,
    this.program,
    this.photoStoragePath,
    this.displayMode = TestimonialDisplayMode.firstNameLastInitial,
    this.showUniversity = true,
    this.showCountry = true,
    this.showProgram = true,
    this.showPhoto = false,
    this.evidenceStoragePaths = const [],
  });

  final String? opportunityId;
  final String opportunityName;
  final String opportunityProvider;
  final String opportunityType;
  final String? country;
  final String? degreeLevel;
  final String? fieldOfStudy;
  final int? successYear;
  final TestimonialOutcome outcome;
  final String? challenge;
  final String? discoveryStory;
  final String? preparationStory;
  final String? scholarsphereHelp;
  final String? outcomeNarrative;
  final String? impact;
  final String? advice;
  final List<String> featuresUsed;
  final String fullName;
  final String? university;
  final String? program;
  final String? photoStoragePath;
  final TestimonialDisplayMode displayMode;
  final bool showUniversity;
  final bool showCountry;
  final bool showProgram;
  final bool showPhoto;
  final List<String> evidenceStoragePaths;

  Map<String, dynamic> toJson() => {
    'opportunity': {
      if (opportunityId != null) 'opportunity_id': opportunityId,
      'opportunity_name': opportunityName,
      'opportunity_provider': opportunityProvider,
      'opportunity_type': opportunityType,
      'country': country,
      'degree_level': degreeLevel,
      'field_of_study': fieldOfStudy,
      'success_year': successYear,
      'outcome': outcome.wire,
    },
    'experience': {
      'challenge': challenge,
      'discovery_story': discoveryStory,
      'preparation_story': preparationStory,
      'scholarsphere_help': scholarsphereHelp,
      'outcome_narrative': outcomeNarrative,
      'impact': impact,
      'advice': advice,
      'features_used': featuresUsed,
    },
    'profile': {
      'full_name': fullName,
      'university': university,
      'program': program,
      'photo_storage_path': photoStoragePath,
    },
    'privacy': {
      'display_mode': displayMode.wire,
      'show_university': showUniversity,
      'show_country': showCountry,
      'show_program': showProgram,
      'show_photo': showPhoto,
    },
    'evidence_storage_paths': evidenceStoragePaths,
  };

  TestimonialDraftInput copyWith({
    String? opportunityId,
    bool clearOpportunityId = false,
    String? opportunityName,
    String? opportunityProvider,
    String? opportunityType,
    String? country,
    String? degreeLevel,
    String? fieldOfStudy,
    int? successYear,
    TestimonialOutcome? outcome,
    String? challenge,
    String? discoveryStory,
    String? preparationStory,
    String? scholarsphereHelp,
    String? outcomeNarrative,
    String? impact,
    String? advice,
    List<String>? featuresUsed,
    String? fullName,
    String? university,
    String? program,
    String? photoStoragePath,
    TestimonialDisplayMode? displayMode,
    bool? showUniversity,
    bool? showCountry,
    bool? showProgram,
    bool? showPhoto,
    List<String>? evidenceStoragePaths,
  }) => TestimonialDraftInput(
    opportunityId: clearOpportunityId
        ? null
        : (opportunityId ?? this.opportunityId),
    opportunityName: opportunityName ?? this.opportunityName,
    opportunityProvider: opportunityProvider ?? this.opportunityProvider,
    opportunityType: opportunityType ?? this.opportunityType,
    country: country ?? this.country,
    degreeLevel: degreeLevel ?? this.degreeLevel,
    fieldOfStudy: fieldOfStudy ?? this.fieldOfStudy,
    successYear: successYear ?? this.successYear,
    outcome: outcome ?? this.outcome,
    challenge: challenge ?? this.challenge,
    discoveryStory: discoveryStory ?? this.discoveryStory,
    preparationStory: preparationStory ?? this.preparationStory,
    scholarsphereHelp: scholarsphereHelp ?? this.scholarsphereHelp,
    outcomeNarrative: outcomeNarrative ?? this.outcomeNarrative,
    impact: impact ?? this.impact,
    advice: advice ?? this.advice,
    featuresUsed: featuresUsed ?? this.featuresUsed,
    fullName: fullName ?? this.fullName,
    university: university ?? this.university,
    program: program ?? this.program,
    photoStoragePath: photoStoragePath ?? this.photoStoragePath,
    displayMode: displayMode ?? this.displayMode,
    showUniversity: showUniversity ?? this.showUniversity,
    showCountry: showCountry ?? this.showCountry,
    showProgram: showProgram ?? this.showProgram,
    showPhoto: showPhoto ?? this.showPhoto,
    evidenceStoragePaths: evidenceStoragePaths ?? this.evidenceStoragePaths,
  );

  static const empty = TestimonialDraftInput(
    opportunityName: '',
    opportunityProvider: '',
    opportunityType: 'scholarship',
    outcome: TestimonialOutcome.awarded,
    fullName: '',
  );
}
